Design document for user customization and persistence

Ebaneck opened the pull request:
**Component**:

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->
'docs', 'design'

**Context**: 

see #2233 

**Summary**:


This document explains some of the design choices considered while
designing a simplistic tool for MetalK8s that guarantees that user-defined
configurations are persisted throughout.


**Acceptance criteria**: 

- Most be validated by Architect and Team Lead

---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #2233

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

bert-e commented:
# Hello ebaneck,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

slaperche-scality commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,214 @@
+  user configurations as listed in the requirements
+- Template and render Deployment and Pod manifests that will make use of
+  this user customizable configurations using default pillar values
+- Provide a CLI tool(Bash script) for changing any of the user configurations:
```
Why Bash instead of Python?
I thought we were moving away from Shell script.

Ebaneck replied:
Very true @slaperche-scality 
I think we have an entire EPIC to handle this migration and it is debatable if we should wait or have something working(shell scripts) while waiting.

Maybe, after all, we do not even have to handle this point within this EPIC

slaperche-scality replied:
The other EPIC is about migrating existing shellscript.

Meanwhile, for new ones that we creates from scratch, it could be a good idea to use Python from the beginning (except if it's really more costly than Bash for some reasons): would be less work for the other EPIC, while adding no extra-work to the current one (anyway you have to write a new script)

gdemonet replied:
In my opinion, wait for the Python CLI epic (#2163) to have the foundations set in place before implementing a CLI for this. In any case, the CLI / UI implementation of those customizations is less urgent than having a framework to persist said customizations over upgrade/downgrade/... IMO, CLI / UI shouldn't be part of the first iteration.

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,214 @@
+Design Choices
+--------------
+
+:term:`ConfigMap` KV(Key-Value) store is chosen as a unified data access and
```
I'm uncomfortable with this focus on "key-value" nature of a `ConfigMap` - does it imply we will use a single `ConfigMap` with a key for each "service" to be configured? I'm not sure that's what we want actually, for multiple reasons (which may be adjusted after we get the Requirements btw):

- CMs are namespace-scoped, and so are most of the services we deploy with MetalK8s. Wouldn't it make more sense if we keep a configuration CM next to what it configures? E.g. `metalk8s-auth-userconfig` in the `metalk8s-auth` namespace?
- Editing a single CM with many keys is less user-friendly than having smaller CMs per service, each with a single key. Although we plan on providing better interfaces through CLI and UI, it would make sense to make the underlying format as friendly as possible (this will make a first iteration with no CLI/GUI much more useful)

Ebaneck replied:
Done, removed the focus on `key-value` nature

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,214 @@
+
+* The limited number of read/write operations even for multi-node setups means
+  that concurrency issues on CRUD operations will not lag even if we do not
+  implement cross-app locking.
```
"concurrency issues [...] will not lag" - not sure what you meant?

In any case, is it important to mention in the design? We're not looking for a database, we're looking for means of configuration. As a matter of fact, we **do have** a database (`etcd`), and a nice schema already in place that we can use (the K8s API resources, including `ConfigMap`s) for such configurations.

The concurrency matters on configuration of cluster-wide services are, as you suggest here, not a problem. Let's move this point to the "non-goals" in the Requirements (cc @thomasdanan), and leave it aside.

We need our design choices to clearly explain **why** we chose this approach according to our "goals", not to our "non-goals", IMO.

Ebaneck replied:
Done

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,214 @@
+* The limited number of read/write operations even for multi-node setups means
+  that concurrency issues on CRUD operations will not lag even if we do not
+  implement cross-app locking.
+* Ability to support CRUD operations on ConfigMap's with CLI and UI easily
```
Why this focus on CRUD? I'm actually thinking, we only want the users to perform **U**pdate operations on those objects, since MetalK8s is responsible for **C**reating and **D**eleting those objects - so the fact that all CRUD ops are supported could be a problem (luckily, we can simply have a `Role` which only allows `patch` on the right objects).

Ebaneck replied:
ok

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,214 @@
+* Ability to support CRUD operations on ConfigMap's with CLI and UI easily
+  using our already existing python kubernetes module.
+* Guarantee of adaptability and ease of changing the design and implementation
+  in cases were customer needs evolve rapidly(modular design nature)
```
What is modular in this design? Especially with the "single ConfigMap KVS" approach.

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,214 @@
+* Guarantee of adaptability and ease of changing the design and implementation
+  in cases were customer needs evolve rapidly(modular design nature)
+* Ease of operation and a guarantee of no downtime in a MetalK8s cluster
+  especially when configuration changes are made on the fly.
```
How does using ConfigMaps handle "no-downtime" guarantees?

gdemonet replied:
> changes are made on the fly.

WDYM? Do you expect a change to impact the target service right away? Or not?
"made on the fly" would mean something like `kubectl edit cm ...` or something else?

Ebaneck replied:
Well, this was supposed to mean; the fact that we are making use of ConfigMaps, I think once a user updates a given configuration, we do not require Pod restarts for the specific services to pick up the changes. no?

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,214 @@
+
+
+Design Choices
+--------------
```
This design only discusses the choice of storage (method and media). What about the concerns of translating these user configurations into services configurations? Are the services supposed to know about our user configs and parse them? Are we (Salt) deriving actual service configs from our user configs? Why Salt and not a K8s Operator?

Ebaneck replied:
I added a `How it works section` that takes care of the questions raised here. This might need validation from You

gdemonet replied:
Good content, though I'd rather state this in the _Design Choices_ section. In addition, there seems to be two implementation alternatives (Salt vs. Operator). Since we apparently decide to use Salt (and not an Operator), why not show Salt as the design + Operator as a rejected option?

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,214 @@
+
+A sample ConfigMap KV store can be defined with the following fields. These
+objects will be namespace bounded ensuring that sensitive information is
+restricted from a cluster-wide access.
```
CMs are always namespace-scoped.

Ebaneck replied:
ok

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,214 @@
+       data:
+         replicas.grafana: 2
+         replicas.prometheus: 2
+         replicas.alertmanager: 1
```
This kind of format is not straightforward to manage with code, since you'll need to parse the flat keys into hierarchical structures. As discussed during the short design session, we'd rather have something more "K8s-like":

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  namespace: metalk8s-monitoring
  name: metalk8s-grafana-userconfig
data:
  config.yaml: |-
    apiVersion: metalk8s.scality.com/v1alpha1
    kind: GrafanaUserConfig
    spec:
      replicas: 2
```

This also has the benefit of bearing a schema version, which we can check before inferring the underlying service configuration.

Ebaneck replied:
Thanks for the reminder! even though i think at the end of the day, we would have littered the entire namespace with many small ConfigMap objects.

What are the chances that we could group the number of replicas specific ConfigMap for all services in the Monitoring namespace?

TeddyAndrieux replied:
Not really agree with your approach here.
If we use a config map IMO we should have just "key value" like Claude did and when we will use specific CRD then we could have object like you suggest.
Also IMO we should do one configmap per "object/service" (e.g.: One configmap for auth/dex, one for grafana, one for prometheus (or maybe one for monitoring regrouping prom, grafana altermanager), one for ...)

Ebaneck replied:
> Thanks for the reminder! even though i think at the end of the day, we would have littered the entire namespace with many small ConfigMap objects.
> 
> What are the chances that we could group the number of replicas specific ConfigMap for all services in the Monitoring namespace?

Oh, let us keep it service-specific as suggested since the configmap of one particular service might grow drastically when new customizations are included

gdemonet replied:
> If we use a config map IMO we should have just "key value" like Claude did and when we will use specific CRD then we could have object like you suggest.

Then it means we'll need to implement the parsing logic twice? Why not stick with the same YAML style/format?

TeddyAndrieux replied:
The configmap will be read from salt so if we use the "kubernetes style" that mean we need to implement it in our salt modules as we will have a configmap with several "key=string" and we will need to then load and parse the string to get the yaml we want.
If we stick with key=value we will just have to read a classic dict in salt so it's really easy.

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,214 @@
+  user configurations as listed in the requirements
+- Template and render Deployment and Pod manifests that will make use of
+  this user customizable configurations using default pillar values
+- Provide a CLI tool(Bash script) for changing any of the user configurations:
```
CLI tooling doesn't need to be part of the first iteration IMO - well formatted CMs are easy enough to manage both manually and in tests/scripts.

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,214 @@
+      staticUser store)
+    - Adding/modifying/removing Alertmanager notification configuration
+
+Iteration 2
```
UI and CLI could both be part of iteration 2 I think, and we could add a third iteration to include pre-edit validation hooks, display of both user config and generated service configs... We'll see with the reqs.

Ebaneck replied:
Moved UI and CLI operations to iteration 2 as suggested. Not sure about third iteration so i'm gonna have to pass!

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,214 @@
+In the Operational Guide:
+
+* Document how to customize or change any given setting using the CLI tool
+* Document how to customize or change any given setting using the UI interface
```
OK, maybe a "common doc" for CLI / UI. But we'll also need this info next to the services we want our users to configure (e.g. show how to configure Dex in the section/page dedicated to Dex)

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,214 @@
+CRUD Operations on the ConfigMap KV store should be tested both from the UI
+test libraries and the CLI tool provided(An initial idea here is to have a
+simplistic configurator.sh bash script(CLI tool) performing sets of CRUD
+operations on a sample ConfigMap KV store).
```
Standard operations to edit configmaps are not super useful to test - what we want to check is propagation of those changes down to the configured services.
E2E tests with CLI and UI are a good thing, because (I guess) those would also trigger the right Salt calls after the CM(s) update(s).

We also need to test edge-cases:
- invalid format for the user configs
- invalid values in the user configs
- lost a user config (someone deleted it)

Ebaneck replied:
ok

gdemonet reviewed (changes requested):
Early review, we'll see how it goes with requirements.
I think you should focus more on "what happens after the user has configured something", not only on "how will the user configure something".

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

TeddyAndrieux commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,225 @@
+  - Apply a salt state that reads the ConfigMap object, validates the schema
+    based on MetalK8s defined standards and checks the new values passed.
+  - If the ConfigMap object is valid, the new values passed by the user are
+    then re-rendered to the pillars.
```
Not sure we really want this value to be in the pillar, IMO just reading the CM directly in the state is better so that user can NOT change, for example, the number of replicas just on CLI using `pillar="blabla"` but he always need to edit the CM

Ebaneck replied:
Fixed

TeddyAndrieux commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,225 @@
+setup this flow(i.e. the act of configuration update by the user and it's
+propagation to the consumers) but the K8s Operator pattern could be use to
+replace configuration synchronization between user defined configurations and
+consumers.
```
Not sure we should talk about it here, for me we should take a desicion in this document and maybe do not implement the operator in the first iteration and only use salt but at least having the final view here is better IMO

TeddyAndrieux commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,225 @@
+      - Extract the pods associated to this ConfigMap based on it's labels
+      - Read and validate the ConfigMap data and schema
+      - If the ConfigMap is valid, update the pillars and restart the
+        respective pods such that they pick-up new configs from the pillars.
```
Not from the pillar in this approach

Ebaneck replied:
As discussed in person... very true!

TeddyAndrieux commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,225 @@
+When using an Operator(a Custom Controller that works with CRDs), we create a
+Custom Resource Definition (CRD) that references a ConfigMap. Once a ConfigMap
+is updated by the user, then the Operator is designed to perform the following
+actions;
```
Operator and CRD are not related we could use CRD with salt and use Operator with ConfigMap.

Also if we use Operator and CRD then we no longer need ConfigMap IMO we directly use CRD as entry point

TeddyAndrieux commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,225 @@
+        respective pods such that they pick-up new configs from the pillars.
+      - If the ConfigMap is invalid, log the error and perform no further
+        action because a bad ConfigMap being applied could lead to cluster
+        outages
```
Hmm I not really like this point, we not directly apply the ConfigMap but we use it to modify another object and i don't see case where an invalid value can cause outages for me apiServer will just not set the wrong value.
For example if you use a string as replica then opeator will try to edit an object and apiServer will just reject the change, ... 
Maybe for other custom tuning that could lead to cluster outages but ... I don't see where

Ebaneck replied:
ok... resolved!

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,224 @@
+  using our already existing python kubernetes module.
+* Guarantee of adaptability and ease of changing the design and implementation
+  in cases were customer needs evolve rapidly.
+* ConfigMap data is store in the :term:`etcd` database which is generally being
```
```suggestion
* ConfigMaps are stored in the :term:`etcd` database which is generally being
```

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,224 @@
+   ConfigMaps.
+   To handle this particular user case, there is a need to provision persistent
+   storage volumes to handle the persistence of these settings across Pod
+   restarts.
```
True. I would create a section for "non-goals" of this design, including this point on Grafana, but also other aspects (providing a general solution for recovery after total site loss, periodic backups, ...).

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,224 @@
+Implementation Details
+----------------------
+
+Storage format
```
The actual format is part of the design. It would be worth adding an explanation in the _Design Choices_ as to why we want to use the traditional versioned `apiVersion/kind` YAML style instead of flat KV style (what benefits it brings, what constraints / extra efforts it requires...).

Ebaneck replied:
Well, we still have an open discussion about the format to use and from what I see, the flat KV style sits well with our already existing python-Kubernetes module....

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,224 @@
+        spec:
+          replicas: 2
+        podSelector:
+          app: grafana
```
Probably missing an indent?

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,224 @@
+
+**Using Saltstates**
+
+Once a ConfigMap KV store is updated by the user(say a user changes the
```
```suggestion
Once a ConfigMap KV store is updated by the user (say a user changes the
```

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,224 @@
+
+During Bootstrap, pillar values for the various services will be pre-filled
+with default values. Service pods and deployments will be configured to either
+consume configuration data directly from their ConfigMap store or fallback to
```
Using simply _ConfigMap_ i.s.o "ConfigMap store" or "ConfigMap KV store" should be sufficient for readers - could you change this here and in other relevant places?

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,224 @@
+consume configuration data directly from their ConfigMap store or fallback to
+the default pillar values if the service specific ConfigMap store was not found.
+
+**Using Saltstates**
```
Let's write this "Salt states" and use the same notation everywhere (no "Saltstate", "salt state", or "salt-state")

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,224 @@
+
+Once a ConfigMap KV store is updated by the user(say a user changes the
+number of replicas for Prometheus deployments to a new value), then perform the
+following actions;
```
```suggestion
following actions:
```

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,224 @@
+    and checks the new values passed.
+
+  - Finally, we make sure that the updated values are picked up by their
+    respective consumers(this might require Pod restarts for changes to take
```
```suggestion
    respective consumers (this might require Pod restarts for changes to take
```

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,224 @@
+
+  - Finally, we make sure that the updated values are picked up by their
+    respective consumers(this might require Pod restarts for changes to take
+    effect). The podSelector field has labels which can be used to detect which
```
There should be no need for us to "detect" what needs to restart: we own these objects, and the configuration options we expose, so it's safe to assume we'll always know what to do after a re-configuration state is applied.

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,224 @@
+because configuration changes are not frequent(for a typical MetalK8s admin,
+changing the number of replicas for a given deployment could happen once in 3
+months) as such, having an operator watch for object changes is not significant
+and not very useful at this point in time.
```
Agreed with your points. Would just add that Operators are especially useful to provide self-healing functionalities on a reactive basis. When a user changes configuration, it is easy to apply (in the same flow) those configurations to the in-cluster objects. Once deployed, those configurations will be handled by the cluster controllers (Deployments, ConfigMap/Secret-based volume mounts, ...), and self-healing comes out of the box. No real need for more than what Salt provides.

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,224 @@
+months) as such, having an operator watch for object changes is not significant
+and not very useful at this point in time.
+
+**Using Operator architecture(Custom Controllers)**
```
No need to provide a design if it's rejected. Simply explain why it's rejected, this will help readers quickly understand the overall architecture document.

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,224 @@
+- Define and deploy new ConfigMap stores that will hold user configurations
+  as listed in the requirements
+- Template and render Deployment and Pod manifests that will make use of
+  this user customizable configurations using pillar values
```
Are we using Pillar values or directly call into `metalk8s_kubernetes` module in the end? cc. @TeddyAndrieux 

Maybe this could be a good fit for `__slots__` (and especially with [dict support in Salt 3000](https://docs.saltstack.com/en/latest/topics/slots/index.html))?

TeddyAndrieux replied:
IMO we should use directly `metalk8s_kubernetes` module.

I think we don't really need `__slots__` here, basic jinja is enough

NicolasT replied:
I'd try to steer away from `__slots__` as well. The values for these settings can be calculated/retrieved before templating starts, so let's do that.

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,224 @@
+  this user customizable configurations using pillar values
+- Document how to change user configurations using kubectl
+- Create and deploy persistent storage volumes for Grafana dashboards and
+  datasources
```
We need to configure Grafana to use a PVC (not create the volumes ourselves). Just do the same as Prom.
Also, if it's out of the design (and I think we have an epic for Grafana: #2218), maybe just don't state it here.

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,224 @@
+  datasources
+- Document how to create these persistent volume for Grafana dashboards and
+  datasources
+- Document how to add Prometheus alerting rules using a ConfigMap. This
```
Why not use the existing `PrometheusRule` CRD that comes with `prometheus-operator`? If it's a conscious decision, it would make sense to explain in the design choices.

Ebaneck replied:
Will keep it simple and resolve to users editing the original CRD!

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,224 @@
+  ruleSelector field, which selects the ConfigMaps from which rules files are
+  loaded from. This means that to add new Prometheus alert rules, Users need to
+  create new ConfigMaps containing alert rules with labels matching what was
+  defined in the Prometheus ruleSelector field.
```
Ah! Actually, this is true if you replace `ConfigMap` with `PrometheusRule`, IIUC. Or maybe not in our version of `prometheus-operator`...? Didn't check

Ebaneck replied:
Not tested, so let's leave it at editing the original CRD.

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,224 @@
+
+- Provide a CLI tool for changing any of the user configurations:
+
+    - Count of replicas for chosen Deployments(Prometheus)
```
General comment: add a space before opening a parenthesis.

```suggestion
    - Count of replicas for chosen Deployments (Prometheus)
```

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,224 @@
+- Provide a CLI tool for changing any of the user configurations:
+
+    - Count of replicas for chosen Deployments(Prometheus)
+    - Updating a Dex authentication connector(OpenLDAP, AD and
```
```suggestion
    - Updating a Dex authentication connector (OpenLDAP, AD and
```

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,224 @@
+    - Updating the Alertmanager notification configuration
+
+- Provide a UI interface that allows Update operations on all user customizable
+  settings based on the above requirements
```
From the requirements, CLI provides this "all settings possible" capability, while GUI provides more user-friendly / less detailed forms to configure certain services (as you described below).
I think (need to be confirmed) this point can be removed.

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,224 @@
+  tool
+* Document how to customize or change any given service settings using the UI
+  interface
+* Document how to add and edit Prometheus alerting rules using the CLI
```
Why this specific example and not all services?

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,224 @@
+- Dex Static User authentication is currently covered in our test-suite and it
+  will make sense to cover atleast one other authentication connector with the
+  easiest being LDAP since we readily have access to OpenLDAP Docker images and
+  automating this process is possible on Scality Cloud
```
Out of scope - this relates to testing Dex, which is part of the Authentication document.

NicolasT commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,199 @@
+During Bootstrap, pillar values for the various services will be pre-filled
+with default values. Service pods and deployments will be configured to either
+consume configuration data directly from their ConfigMap or fallback to
+the default pillar values if the service specific ConfigMap was not found.
```
What if we did this differently?
None of these 'user configurable' settings (at this level) should influence how the K8s cluster behaves: this should be dictated by the `BootstrapConfiguration`. As such during bootstrap, at some point *before we need any of these settings*, the K8s cluster is fully available.

Which means that at this point, we could create all these `ConfigMap`s, filled with whatever default values we choose.

At this point, we can consume them while deploying various services without any more 'defaults' defined anywhere.

Ebaneck replied:
True! after a couple of discussions, an alternative we considered and updated the design docs was keeping these 'defaults' in our `default.yaml` file and then doing something like ```{%- from "metalk8s/map.jinja" import addons with context %}```. This meant in our add-ons charts we had to implement some conditional to read from the Configmaps or fall back to these defined defaults(just like pillar pre-filling idea but cleaner).

Thanks to your approach, it appears we do not even need to declare defaults in `defaults.yaml`.  Waiting to hear from others what they think.

TeddyAndrieux replied:
Like I told you Monday 3 choice:
- We have `defaults.yaml` containing this default number of replica
- we create the custos CM and "pre-fill" them **if they do not exists**
- We just put default value directly in the `sls` file

Personally suggest the second one as it will be much simpler for the user to change them if CM already exists and already contain value we want

gdemonet replied:
Option 2 looks good to me as well. A simple `kubectl delete cm ..` followed by `salt-run state.orch metalk8s.deployed` to restore defaults is acceptable I guess.

NicolasT commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,199 @@
+
+  - Finally, we make sure that the updated values are picked up by their
+    respective consumers (this might require Pod restarts for changes to take
+    effect).
```
The fact that Pods may be restarted is always the case in K8s, not sure this should be mentioned. In the end, Salt (MetalK8s) is not restarting any Pods. This should all be managed by the `Deployment`/`StatefulSet`/`DaemonSet` controllers because Salt updates some high-level object(s).

Ebaneck replied:
removed

NicolasT commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,199 @@
+
+Note that, Salt states are used to sync data and update consumers
+of new configurations changes mainly because of the minimum effort it takes to
+setup this flow but the K8s Operator pattern could be use to replace
```
```suggestion
setup this flow but the K8s Operator pattern could be used to replace
```

NicolasT commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,199 @@
+Note that, Salt states are used to sync data and update consumers
+of new configurations changes mainly because of the minimum effort it takes to
+setup this flow but the K8s Operator pattern could be use to replace
+configuration synchronization between user defined configurations and
```
`... could be used to synchronize configuration between user-defined settings and the applications.`

NicolasT commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,199 @@
+- Persisting newly added Grafana dashboards or new Grafana datasources
+  especially for modifications added via the Grafana UI cannot be stored in
+  ConfigMaps.
+  To handle this particular user case, we need to configure Grafana deployments
```
```suggestion
  To handle this particular use-case, we need to configure Grafana deployments
```

NicolasT commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,199 @@
+  especially for modifications added via the Grafana UI cannot be stored in
+  ConfigMaps.
+  To handle this particular user case, we need to configure Grafana deployments
+  to use persistent volume claims and then document how to create volumes that
```
Here you're prescribing *how* a certain feature *not* covered by this design ought to be implemented. That's not OK. Indeed, this document should describe what it *won't* cover, but whether this other feature is implemented using PVs or some other mechanism is to be determined later/elsewhere.

Ebaneck replied:
True, I will move it to non-goals and clearly state that the current design does not cater for this. No need to mention an implementation choice.

NicolasT commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,199 @@
+  will handle the persistence of these settings across Pod restarts.
+
+- ConfigMaps as used in the above design for storing and persisting user
+  settings need periodic backups for recovery purposes. This implies that
```
I get what you mean but I don't like how it's phrased. We don't need to put backups in place to store these user settings and such. Instead, using `ConfigMap`s (or other K8s object types) has as a benefit that K8s cluster backups include these settings. The fact that the K8s `etcd` cluster needs to be backed up is not driven by this feature.

NicolasT commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,199 @@
+much more effort to realize and there is no real need for applying changes
+using this method because configuration changes are not frequent
+(for a typical MetalK8s admin, changing the number of replicas for a given
+deployment could happen once in 3 months) as such, having an operator watch for
```
Or less.

NicolasT commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,199 @@
+    data:
+      config.yaml: |-
+        apiVersion: metalk8s.scality.com/v1alpha1
+        kind: GrafanaUserConfig
```
Not sure I like the `GrafanaUserConfig` name. We're not configuring Grafana users here... `GrafanaCustomizations`? `GrafanaConfig`? ...

Ebaneck replied:
Picked the latter!

NicolasT commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,199 @@
+        apiVersion: metalk8s.scality.com/v1alpha1
+        kind: GrafanaUserConfig
+        spec:
+          replicas: 2
```
This is a low-level setting at top level in the config. Not sure that's great.
```yaml
spec:
  deployment:
    replicas: 2
```
(then other `deployment`-related settings can go there as well)

Ebaneck replied:
ok

NicolasT commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,199 @@
+- Template and render Deployment and Pod manifests that will make use of
+  this user customizable configurations using pillar values
+- Document how to change user configurations using kubectl
+- Document how to add Prometheus alerting rules by editing the Prometheus rule
```
I don't see how this is in scope of this design?

Ebaneck replied:
fixed

NicolasT commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,199 @@
+- Provide a UI interface for changing the number of replicas for a chosen set
+  of MetalK8s deployments (Prometheus, ....)
+- Add a UI interface for listing Alertmanager notification systems MetalK8s
+  will support (Slack, email, hipchat)
```
HipChat is dead, no need to list it anywhere.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @gdemonet

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/add-design-documentat-for-user-customization_and_persistence` into
`development/2.5`
* [w/2.6/feature/add-design-documentat-for-user-customization_and_persistence](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/feature/add-design-documentat-for-user-customization_and_persistence) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @gdemonet

bert-e commented:
# History mismatch



Merge commit #d22dc3ee8109149939ce67dc1a804b9d7def75f7 on the integration branch
`w/2.6/feature/add-design-documentat-for-user-customization_and_persistence` is merging a branch which is neither the current
branch `feature/add-design-documentat-for-user-customization_and_persistence` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `feature/add-design-documentat-for-user-customization_and_persistence` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,187 @@
+performing full backups, system restores for a full fledged KV system
+requires time and much more efforts than the ConfigMap approach.
+
+Operator (Custom Controller) vs ConfigMap
```
`vs. Salt` - as Teddy said, Operator is not related to a given storage medium, and could work with plain CM

gdemonet commented on docs/developer/architecture/user_customization_and_persistence.rst:
```diff
@@ -0,0 +1,187 @@
+Storage format
+~~~~~~~~~~~~~~
+
+A sample ConfigMap can be defined with the following fields.
```
Why this format? Explain in the design, compare with flat KV approach and why we prefer this one.

Ebaneck replied:
Done

Ebaneck committed:
doc: Add a design doc for the persistence of cluster and service configurations

Initially, user defined configurations such as changing the number of
replicas for a given deployment in a MetalK8s cluster are lost during
an upgrade/downgrade scenario.

This document explains some of the design choices considered while
designing a simplistic tool for MetalK8s that guarantees that user defined
configurations are persisted throughout.

Closes: #2233

bert-e commented:
# History mismatch



Merge commit #7c2fc3881cb2a9af6100aa8bc127eade1d2983d7 on the integration branch
`w/2.6/feature/add-design-documentat-for-user-customization_and_persistence` is merging a branch which is neither the current
branch `feature/add-design-documentat-for-user-customization_and_persistence` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `feature/add-design-documentat-for-user-customization_and_persistence` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Ebaneck commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/add-design-documentat-for-user-customization_and_persistence` into
`development/2.5`
* [w/2.6/feature/add-design-documentat-for-user-customization_and_persistence](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/feature/add-design-documentat-for-user-customization_and_persistence) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @gdemonet

Ebaneck commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @gdemonet







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @gdemonet







*The following options are set:* **approve**

gdemonet commented on docs/developer/architecture/configurations.rst:
```diff
@@ -73,3 +73,205 @@ As a MetalK8s expert, I can use ``kubectl`` command(s) in order to edit all
+cluster and service configurations in a MetalK8s cluster based on the above
+requirements for the following reasons:
+
+* Ability to support Update operations on ConfigMap's with CLI and UI easily
```
```suggestion
* Ability to support Update operations on ConfigMaps with CLI and UI easily
```

gdemonet commented on docs/developer/architecture/configurations.rst:
```diff
@@ -73,3 +73,205 @@ As a MetalK8s expert, I can use ``kubectl`` command(s) in order to edit all
+* Ability to support Update operations on ConfigMap's with CLI and UI easily
+  using our already existing python kubernetes module.
+* Guarantee of adaptability and ease of changing the design and implementation
+  in cases were customer needs evolve rapidly.
```
```suggestion
  in cases where customer needs evolve rapidly.
```

gdemonet commented on docs/developer/architecture/configurations.rst:
```diff
@@ -73,3 +73,205 @@ As a MetalK8s expert, I can use ``kubectl`` command(s) in order to edit all
+BootstrapConfiguration dictates the cluster behaviour. Also the creation of
+cluster and service configurations takes place before the respective consumers
+(MetalK8s addons) are deployed hence configuration data is always ready for
+consumers and not the other way round.
```
I think you're trying to say that service configurations (and maybe others that affect the cluster **after** initial deployment) are a good fit for ConfigMaps, whereas BootstrapConfiguration needs to stay as a file, since we need it before initial deployment. It also evokes the need for us to create said CMs before deploying the configured services.

Rewording a bit, we could simply state that:
- we need special casing for some configs (what goes in BootstrapConfig)
- we can store all the rest in the API, as long as we make sure these CMs are available **before** we deploy the configured services

Also:
> cluster [...] configurations do not influence how the cluster behaves

This looks incorrect, or if it's not, then I guess those configurations are useless? :thinking:

gdemonet commented on docs/developer/architecture/configurations.rst:
```diff
@@ -73,3 +73,205 @@ As a MetalK8s expert, I can use ``kubectl`` command(s) in order to edit all
+Storage format
+~~~~~~~~~~~~~~
+
+A YAML(K8s-like) format was chosen to represent the data field instead of a
```
```suggestion
A YAML (K8s-like) format was chosen to represent the data field instead of a
```

gdemonet commented on docs/developer/architecture/configurations.rst:
```diff
@@ -73,3 +73,205 @@ As a MetalK8s expert, I can use ``kubectl`` command(s) in order to edit all
+ - YAML formatted configurations are easy to write and understand hence it will
+   be simpler for users to edit configurations.
+ - The YAML format benefits from bearing a schema version, which can be checked
+   before inferring the underlying service configuration if need arises.
```
> checked before inferring

What do you mean? We don't want to hold default values or infer anything on our side. We just want to validate that the "version" of the configuration matches what we deploy.

gdemonet commented on docs/developer/architecture/configurations.rst:
```diff
@@ -73,3 +73,205 @@ As a MetalK8s expert, I can use ``kubectl`` command(s) in order to edit all
+ - The YAML format utilizes white spaces to define variable hierarchies and as
+   such it is easier to manage with code mean while the flat key-value format
+   requires parsing of flat keys into hierarchical structures which can be
+   tedious.
```
True, but not super clear. Maybe simply state that YAML **is** a format for describing hierarchical data structures, while using a flat structure would require a form of encoding (and then, decoding) of this hierarchical structure, which we know we'll need.

gdemonet commented on docs/developer/architecture/configurations.rst:
```diff
@@ -73,3 +73,205 @@ As a MetalK8s expert, I can use ``kubectl`` command(s) in order to edit all
+
+- As stated in the requirements, adding and editing Prometheus alert rules
+  is also not covered by the chosen design choice and will need to be addressed
+  differently.
```
Not clear enough why we don't cover Prometheus rules (and not only alerting rules are affected, btw). Explain that even if we **could** use ConfigMaps for said rules, we **prefer** relying on the Prometheus Operator (which we deploy) and its CRDs (especially, the `PrometheusRule` one).

Ebaneck replied:
As earlier indicated by Nicolas, these are cases that our design does not cater for and it is not required state the solution we want to use.

gdemonet commented on docs/developer/architecture/configurations.rst:
```diff
@@ -73,3 +73,205 @@ As a MetalK8s expert, I can use ``kubectl`` command(s) in order to edit all
+performing full backups, system restores for a full fledged KV system
+requires time and much more efforts than the ConfigMap approach.
+
+Operator (Custom Controller) vs Salt
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Operators are useful in that, they provide self-healing functionalities on a
+reactive basis. When a user changes a given configuration, it is easy to
+reconcile and apply these changes to the in-cluster objects.
+
+The Operator approach was rejected because it is much more complex, requires
+much more effort to realize and there is no real need for applying changes
+using this method because configuration changes are not frequent
+(for a typical MetalK8s admin, changing the number of replicas for a given
+deployment could happen once in 3 months or less) as such, having an operator
+watch for object changes is not significant and not very useful at this point
+in time.
```
Good explanation. Maybe add a small paragraph explaining that we will design Salt formulas to be idempotent (as we do for all formulas), making it easy to apply after a config change and "reconcile the desired state", and just loose the "reactive" aspect of it by making this design choice.

Ebaneck replied:
Config changes in this design and our current approach are not handled by salt I think...it is a Kubernetes thing right?? (kubectl edit). Also, state reconciliation is done by Kubernetes.. no??

gdemonet replied:
Sorry if I wasn't clear. In the current approach, we have:

- Salt creates ConfigMaps for all services with default values during bootstrap
- Salt deploys services with templated values coming from those default ConfigMaps
- Admin sets the desired configurations by `kubectl edit cm ...` if required (or through CLI/UI)

_At that time, the deployed services are still using default values_

- Admin requests (which a CLI/UI could do after editing the ConfigMap) a re-deployment of the desired service, for Salt to orchestrate

This implies a two-step process (whereas the Operator approach only requires the CM edition to have changes applied), *unless* we implement a reactive behaviour (through Beacon+Reactor pairs) directly in Salt master(s). _(this makes for a nice benefit of ConfigMaps vs. CRDs, as the former can be mounted as files, and reacting to file changes is rather easy with Salt)_

TeddyAndrieux replied:
Personnaly prefer to have operator instead of beacon + reactor for this (even if for first iteration having a two-step process with salt is better/simpler IMO)

gdemonet replied:
Well, we can easily transform Helm charts into Operators using Operator SDK, could be worth looking into.

gdemonet commented on docs/developer/architecture/configurations.rst:
```diff
@@ -73,3 +73,205 @@ As a MetalK8s expert, I can use ``kubectl`` command(s) in order to edit all
+^^^^^^^^^^^
+
+- Define and deploy new ConfigMap stores that will hold cluster and service
+  configurations as listed in the requirements
```
Indeed. To provide a bit more details, maybe add:

> for each CM, define its schema, its default values, and how it impacts the configured services

gdemonet commented on docs/developer/architecture/configurations.rst:
```diff
@@ -73,3 +73,205 @@ As a MetalK8s expert, I can use ``kubectl`` command(s) in order to edit all
+- Provide a UI interface for enabling or disabling Dex authentication
+  connectors (LDAP, Active Directory, StaticUser store)
+- Provide a UI interface for changing the number of replicas for a chosen set
+  of MetalK8s deployments (Prometheus, ....)
```
```suggestion
  of MetalK8s deployments (Prometheus, ...)
```

gdemonet commented on docs/developer/architecture/configurations.rst:
```diff
@@ -73,3 +73,205 @@ As a MetalK8s expert, I can use ``kubectl`` command(s) in order to edit all
+  connectors (LDAP, Active Directory, StaticUser store)
+- Provide a UI interface for changing the number of replicas for a chosen set
+  of MetalK8s deployments (Prometheus, ....)
+- Add a UI interface for listing Alertmanager notification systems MetalK8s
```
It's not that "MetalK8s will support", we want to allow anything Alertmanager supports, IMO.

The comment from Nicolas was just that HipChat is EOL, so no need to mention it as an example (no one will use it).

Ebaneck replied:
We are talking about 9 different Alertmanager channels for notifications and i am not sure we want to support all those IMHO.

So, i prefer to leave it generic

gdemonet reviewed (commented):
Almost perfect! Good job ;)

bert-e commented:
# History mismatch



Merge commit #899106f843ef62a9a94dfa50779fda0273a03b0e on the integration branch
`w/2.6/feature/add-design-documentat-for-user-customization_and_persistence` is merging a branch which is neither the current
branch `feature/add-design-documentat-for-user-customization_and_persistence` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `feature/add-design-documentat-for-user-customization_and_persistence` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

thomasdanan commented on docs/developer/architecture/configurations.rst:
```diff
@@ -73,3 +73,197 @@ As a MetalK8s expert, I can use ``kubectl`` command(s) in order to edit all
+
+  - Apply a Salt states that reads the ConfigMap object, validates the schema
+    and checks the new values passed.
+
```
@Ebaneck so If I understand well when we want to change one CSC, manually, it will be done in 2 steps: changing the conf and then applying the conf right?
I guess for some configurations, that may happen frequently (like setting alert notification), a UI or a CLI will automate those 2 steps?

gdemonet replied:
> may happen frequently (like setting alert notification)

"Frequently" here is still closer to monthly than daily IMO, and so gathering both steps into one shouldn't be a top priority. That being said, it's entirely achievable to have those grouped in a single "action", be it in CLI (once #2163 is done) or GUI.

thomasdanan commented on docs/developer/architecture/configurations.rst:
```diff
@@ -73,3 +73,197 @@ As a MetalK8s expert, I can use ``kubectl`` command(s) in order to edit all
+
+- Persisting newly added Grafana dashboards or new Grafana datasources
+  especially for modifications added via the Grafana UI cannot be stored in
+  ConfigMaps and hence will be catered for later.
```
So what will be the solution for this?

Ebaneck replied:
The quickest approach will be to provision persistent volumes for Grafana just like we did for Prometheus & Alertmanager. The effort of doing this is really minimal.

gdemonet replied:
> The effort of doing this is really minimal.

Adding a PV is indeed no effort on our side. However, be it for Grafana or the rest of our services which use persistent storage, I'm not sure how we'll handle upgrade (and, if required, migration of their data). This question may not be as trivial to solve as "adding a PV".

thomasdanan commented on docs/developer/architecture/configurations.rst:
```diff
@@ -73,3 +73,197 @@ As a MetalK8s expert, I can use ``kubectl`` command(s) in order to edit all
+- As stated in the requirements, adding and editing Prometheus alert rules
+  is also not covered by the chosen design choice and will need to be addressed
+  differently. Even if we could use ConfigMaps for Prometheus rules, we prefer
+  relying on the Prometheus Operator and it's CRD (PrometheusRule).
```
Does it mean someone wanting to change the threshold of one rule can do it but will lose its configuration after an Upgrade? what solution do we suggest for this specific case? Is alert notification configuration is one of the things that is not supported by this design?

Ebaneck replied:
Alertmanager notification configurations are handled in our current design since we can store whatever notification configuration ( e.g slack, email) we wish to implement in ConfigMaps. PS We have a ticket for storing the Alertmanager notification configurations.

For changing those thresholds, we found out that you can actually do it in the original PrometheusRule CRD deployed by the Prometheus-operator and I want to believe the Crds are persisted even after upgrades.
We just need to provide adequate documentation explaining how this can be done.

Remember that this design doc is about using ConfigMaps to store and persist Service & Cluster configurations. Persisting or changing Prometheus rules is not handled using ConfigMaps and hence we were advised not to add that particular solution in this document.

gdemonet replied:
> do it in the original PrometheusRule CRD deployed by the Prometheus-operator and I want to believe the Crds are persisted even after upgrades.

CRDs belong to the Operator - a cluster admin could add rules by creating `PrometheusRule` objects, as he would add volumes by creating `Volume` objects.
Those *new* objects are not known by us, so we will leave them untouched under all circumstances.

> we were advised not to add that particular solution in this document.

At least in this first design PR. It may make sense to group all possible solutions to the general "configuration" problem under a single umbrella, but for now, we focus on what's missing: said ConfigMaps.

thomasdanan commented on docs/developer/architecture/configurations.rst:
```diff
@@ -73,3 +73,197 @@ As a MetalK8s expert, I can use ``kubectl`` command(s) in order to edit all
+* Document how to customize or change any given service settings using the CLI
+  tool
+* Document how to customize or change any given service settings using the UI
+  interface
```
So If it is supported through a UI, can we say it is supported through the API? (and not only CLI). Again if changing configuration is done in 2 steps, first step would be through K8s API, to change config map and second one will be through Salt API to apply the change?

gdemonet replied:
Indeed, v1 of the docs should document the ConfigMaps, how to edit them (`kubectl edit` / `patch` or directly using the HTTP API), and how to apply changes using Salt.

Ebaneck replied:
True and should be straight forward to handle in the docs
E.g

i. Configmap: metalk8s-dex-config

ii. How to edit it?

`kubectl edit configmaps metalk8s-dex-config -n metalk8s-auth`

ii. Apply the changes to take effect?

`salt-run state.sls metalk8s.addons.dex.deployed.chart saltenv=metalk8s-2.5.0-dev`

`kubectl rollout restart deployment dex -n metalk8s-auth`

TeddyAndrieux replied:
> kubectl rollout restart deployment dex -n metalk8s-auth

Why ?

Ebaneck replied:
Well, it looks like we need to forcefully restart the deployments...

Switching my dex replicas count from (2) to (1) and then applying the state didn't seem to change anything until I restarted the deployment.

thomasdanan commented on docs/developer/architecture/configurations.rst:
```diff
@@ -73,3 +73,197 @@ As a MetalK8s expert, I can use ``kubectl`` command(s) in order to edit all
+  interface
+* Document the list of service settings which can be configured by the user
+
+Test Plan
```
Shouldn't you check that an upgrade of MetalK8s doesn't break previous customizations?

Ebaneck replied:
Indeed, we should

Ebaneck commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/add-design-documentat-for-user-customization_and_persistence` into
`development/2.5`
* [w/2.6/feature/add-design-documentat-for-user-customization_and_persistence](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/feature/add-design-documentat-for-user-customization_and_persistence) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @gdemonet







*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #8407ee2e7c57f212bf60343a3111372f21726e0e on the integration branch
`w/2.6/feature/add-design-documentat-for-user-customization_and_persistence` is merging a branch which is neither the current
branch `feature/add-design-documentat-for-user-customization_and_persistence` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `feature/add-design-documentat-for-user-customization_and_persistence` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

Ebaneck commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

Ebaneck commented:
/reset

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/add-design-documentat-for-user-customization_and_persistence` into
`development/2.5`
* [w/2.6/feature/add-design-documentat-for-user-customization_and_persistence](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/feature/add-design-documentat-for-user-customization_and_persistence) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @gdemonet







*The following options are set:* **approve**

Ebaneck commented:
/status

bert-e commented:
# Status





*Status report is not available.*






*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @gdemonet







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @gdemonet







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet







*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

Ebaneck commented:
/help

bert-e commented:
# Help page



The following options and commands are available at this time.

## Options ##


name   | description  | privileged   | authored
------ | ------------ | ------------ |----------
:arrow_right: **create_pull_requests** | Allow the creation of integration pull requests. |  | 
:arrow_right: **bypass_author_approval** | Bypass the pull request author's approval |  :star:  | 
:arrow_right: **bypass_commit_size** | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
:arrow_right: **unanimity** | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
:arrow_right: **bypass_incompatible_branch** | Bypass the check on the source branch prefix |  :star:  | 
:arrow_right: **bypass_build_status** | Bypass the build and test status |  :star:  | 
:arrow_right: **approve** | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
:arrow_right: **after_pull_request** | Wait for the given pull request id to be merged before continuing with the current one. |  | 
:arrow_right: **wait** | Instruct Bert-E not to run until further notice. |  | 
:arrow_right: **bypass_jira_check** | Bypass the Jira issue check |  :star:  | 
:arrow_right: **no_octopus** | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
:arrow_right: **bypass_leader_approval** | Bypass the pull request leaders' approval |  :star:  | 
:arrow_right: **bypass_peer_approval** | Bypass the pull request peers' approval |  :star:  | 



## Commands ##


name   | description  | privileged
------ | ------------ | ------------
:arrow_right: **reset** | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
:arrow_right: **help** | Print Bert-E's manual in the pull request. | 
:arrow_right: **build** | Re-start a fresh build ```TBA``` | 
:arrow_right: **force_reset** | Delete integration branches & pull requests, and restart merge process from the beginning. | 
:arrow_right: **retry** | Re-start a fresh build ```TBA``` | 
:arrow_right: **status** | Print Bert-E's current status in the pull request ```TBA``` | 
:arrow_right: **clear** | Remove all comments from Bert-E from the history ```TBA``` | 






*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



Please check the status of the associated issue None.

Goodbye ebaneck.
